File: procedural-languages/Scala/caesar-cipher.scala
object CaesarCipher {
  def main(args: Array[String]): Unit = {
    print("Enter text to encrypt: ")
    val inputEncrypt = scala.io.StdIn.readLine()
    print("Enter shift value: ")
    val shiftEncrypt = scala.io.StdIn.readInt()
    println("Encrypted text: " + encrypt(inputEncrypt, shiftEncrypt))

    print("Enter text to decrypt: ")
    val inputDecrypt = scala.io.StdIn.readLine()
    print("Enter shift value: ")
    val shiftDecrypt = scala.io.StdIn.readInt()
    println("Decrypted text: " + decrypt(inputDecrypt, shiftDecrypt))

    print("Enter text for brute-force solve: ")
    val inputSolve = scala.io.StdIn.readLine()
    solve(inputSolve)
  }

  def encrypt(input: String, shift: Int): String = {
    val upperInput = input.toUpperCase
    val encrypted = new StringBuilder

    for (char <- upperInput) {
      if (char.isLetter) {
        val shifted = ((char - 'A' + shift) % 26 + 'A').toChar
        encrypted.append(shifted)
      } else {
        encrypted.append(char)
      }
    }
    encrypted.toString()
  }

  def decrypt(input: String, shift: Int): String = {
    encrypt(input, 26 - (shift % 26))
  }

  def solve(input: String): Unit = {
    println("Solving the cipher with all 26 possible shifts:")
    for (shift <- 1 to 26) {
      println(s"Shift $shift: ${decrypt(input, shift)}")
    }
  }
}

File: functional-languages/ML/caesar-cipher.ml
(* Shift a single character by n positions *)
fun shiftChar (c: char, shift: int) : char =
  let
    val code = ord c
    val baseUpper = ord #"A"
    val baseLower = ord #"a"
    fun mod26 n = if n < 0 then mod26 (n + 26) else n mod 26
  in
    if code >= ord #"A" andalso code <= ord #"Z" then
      chr (baseUpper + mod26 (code - baseUpper + shift))
    else if code >= ord #"a" andalso code <= ord #"z" then
      chr (baseLower + mod26 (code - baseLower + shift))
    else
      c
  end

(* Encrypt a string by shifting each character *)
fun encrypt (text: string, shift: int) : string =
  String.implode (List.map (fn c => shiftChar (c, shift)) (String.explode text))

(* Decrypt a string by shifting in the opposite direction *)
fun decrypt (text: string, shift: int) : string =
  encrypt (text, ~shift)

(* Brute force all possible Caesar shifts *)
fun bruteForce (text: string) : unit =
  let
    fun tryShifts n =
      if n > 26 then ()
      else (
        print ("Shift " ^ Int.toString n ^ ": " ^ encrypt(text, 26 - n) ^ "\n");
        tryShifts (n + 1)
      )
  in
    tryShifts 1
  end

(* Example I/O interaction *)
val () =
  let
    val _ = print "Enter text to encrypt: "
    val inputEnc = TextIO.inputLine TextIO.stdIn |> valOf |> String.trim
    val _ = print "Enter shift value: "
    val shiftEnc = TextIO.inputLine TextIO.stdIn |> valOf |> Int.fromString |> valOf
    val encrypted = encrypt(inputEnc, shiftEnc)
    val _ = print ("Encrypted text: " ^ encrypted ^ "\n\n")

    val _ = print "Enter text to decrypt: "
    val inputDec = TextIO.inputLine TextIO.stdIn |> valOf |> String.trim
    val _ = print "Enter shift value: "
    val shiftDec = TextIO.inputLine TextIO.stdIn |> valOf |> Int.fromString |> valOf
    val decrypted = decrypt(inputDec, shiftDec)
    val _ = print ("Decrypted text: " ^ decrypted ^ "\n\n")

    val _ = print "Enter text for brute-force solve: "
    val inputBrute = TextIO.inputLine TextIO.stdIn |> valOf |> String.trim
    val _ = print "Solving the cipher with all 26 possible shifts:\n"
    val _ = bruteForce inputBrute
  in
    ()
  end